plugin/path_impersonated_account_secrets.go

```diff
@@ -45,10 +58,28 @@ func (b *backend) pathImpersonatedAccountAccessToken(ctx context.Context, req *l
 		return nil, err
 	}
 
+	config, err := getConfig(ctx, req.Storage)
+	if err != nil {
+		return nil, err
+	}
+
 	tokenSource, err := impersonate.CredentialsTokenSource(ctx, impersonate.CredentialsConfig{
 		TargetPrincipal: acct.EmailOrId,
 		Scopes:          acct.TokenScopes,
+		Lifetime:        config.MaxTTL,
```

Comment:
maybe it should be `config.TTL` instead since that would be the expected default TTL for the mount. I'm not sure if each role can also have its own TTL set. If they can then I think we should try to use that and if it doesn't exist fallback to the mount default TTL.

Reply:
I'm happy to make that change.  So in our case we'd configure this so that `TTL` and `MaxTTL` we both set to `4h` in our case?

Reply:
yep

Reply:
> I'm not sure if each role can also have its own TTL set. If they can then I think we should try to use that and if it doesn't exist fallback to the mount default TTL. That way we could possibly lower the TTL per role if that is required in the future.

Big +1 on each role having its own TTL because that is more in-line with how rolesets get configured. And the role configured TTL should be within the plugin configured TTL and maxTTL. And the fallback should use the configured plugin configured TTL.

Note that currently, the plugin configured TTL and maxTTL are both `0s`.

Reply:
> Big +1 on each role having its own TTL because that is more in-line with how rolesets get configured

I haven't uncovered any documentation in GCP that allows this?  

> maybe it should be config.TTL instead since that would be the expected default TTL for the mount. 

I have some reservations about raising the Minimum `config.TTL` to 4 hours and using that.  I get concerned when I see a minimum value raised beyond what is allowable as a default by GCP.  This is subjective, but I find that having a minimum value specified that works for any GCP project and a Maximum value specified that *might* work if the project is configured in a certain way, is safer and more intuitive.  If I change this to:

```go
Lifetime: config.TTL
```

This will require setting the `TTL` to `4hr` which might fail of the project configuration changes or is done improperly sometime.

Reply:
Just got off the Zoom call with Brandon, Brian, and Ian. We are going to go with the following:

- Update the Impersonated Account role with a TTL field
- Use the mount configuration maxTTL as a validation to ensure that the Impersonated Account role TTL is not greater. This should be set to 12h.
- In the event that the Impersonated Account role TTL field is not set, the GCP access token TTL default of 1hr should be used. The mount configuration TTL should be set to 1hr so that it is consistent with AWS.
- vault-binding-controller will be updated to configured the Impersonated Account TTL field to 4hr.

Reply:
It's probably also worth pointing out that if we make `ttl` a parameter on `impersonated-account`:

https://github.com/DataDog/vault-plugin-secrets-gcp/blob/6d758da3e973655622395dfd8a2f98be4a84ad98/plugin/path_impersonated_account.go#L30-L33

Rather than a `ttl` parameter when reading the secret on `*/impersonated-account/token`:

https://github.com/DataDog/vault-plugin-secrets-gcp/blob/6d758da3e973655622395dfd8a2f98be4a84ad98/plugin/path_impersonated_account_secrets.go#L18-L21

This differs a bit from where the `ttl` is specified when requesting a secret on `*/static-account/key`:

https://github.com/DataDog/vault-plugin-secrets-gcp/blob/6d758da3e973655622395dfd8a2f98be4a84ad98/plugin/path_static_account_secrets.go#L29-L32

The only reason I bring this up is that since the `/impersonated-account/token` path is read-only, we don't have to worry about possibly creating data where the `ttl` on the impersonated account is `12h` but the mount config sets a `maxTTL` to `4h`.  

So Question: If the `maxTTL` on the Mount is 4h, should we return an error if the `ttl` on the `impersonated-account` is `12h`?

Reply:
I don't think we should return an error in that case. I think the aws plugin truncate the ttl to the max for the mount and then adds a warning to the response.

Reply:
Oh interesting.  In that case VBC might keep making the same change then realizing the current config != desired config 🤔

Reply:
oh, when the role is configured the ttl isn't truncated. It is only when the ttl is going to be used to retrieve credentials.

So if vault binding controller configures a role with a ttl of 28 hours even though the max is set to 4 hours then when the role is read back it will have a ttl of 28 hours still. When credentials are retrieved for that role the ttl is truncated to 4 hours and a warning is added to the response.